File: src/main/java/com/teamwizardry/librarianlib/api/gui/Option.java
package com.teamwizardry.librarianlib.api.gui;

import java.util.function.Function;

/**
 * An option that can be defined by setting a value or by a callback
 * @author Pierce Corcoran
 *
 * @param <T> The type returned
 * @param <P> The type of the parameter to the option
 */
public class Option<P, T> {

	protected T value;
	protected T defaultValue;
	protected Function<P, T> callback;
	
	public Option(T defaultValue) {
		this.defaultValue = defaultValue;
	}
	
	public T getValue(P param) {
		if(callback == null)
			return value == null ? defaultValue : value;
		return callback.apply(param);
	}
	
	public void func(Function<P, T> callback) {
		this.callback = callback;
	}
	
	public void noFunc() {
		this.callback = null;
	}
	
	public void setValue(T value) {
		this.value = value;
	}
	
}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/Option.kt
package com.teamwizardry.librarianlib.client.gui

import java.util.function.Function

/**
 * An option that can be defined by setting a value or by a callback
 * @author Pierce Corcoran
 *
 *
 * @param T The type returned
 *
 * @param P The type of the parameter to the option
 */
open class Option<P, T>(protected var defaultValue: T) {

    private var value: T = defaultValue
    protected var callback: Function<P, T>? = null

    fun getValue(param: P): T {
        val tmp = callback
        if (tmp != null) {
            return tmp.apply(param)
        }
        return value
    }

    fun func(callback: ((P) -> T)?) {
        if (callback == null)
            func(null as Function<P, T>?)
        else
            func(Function<P, T>(callback))
    }

    fun func(callback: Function<P, T>?) {
        this.callback = callback
    }

    fun noFunc() {
        this.callback = null
    }

    fun setValue(value: T) {
        this.value = value
    }

    operator fun invoke(p: P): T {
        return getValue(p)
    }

    operator fun invoke(t: T) {
        setValue(t)
    }

    operator fun invoke(callback: ((P) -> T)?) {
        func(callback)
    }

}
